examples/tv-casting-app/android/App/app/src/main/jni/cpp/core/CastingPlayer-JNI.cpp

```diff
@@ -0,0 +1,152 @@
+#include <app/clusters/bindings/BindingManager.h>
+#include <app/server/Server.h>
+#include <jni.h>
+#include <lib/support/JniReferences.h>
+#include <lib/support/JniTypeWrappers.h>
+
+using namespace chip;
+
+#define JNI_METHOD(RETURN, METHOD_NAME)                                                                                            \
+    extern "C" JNIEXPORT RETURN JNICALL Java_com_matter_casting_core_MatterCastingPlayer_##METHOD_NAME
+
+namespace matter {
+namespace casting {
+namespace core {
+
+JNI_METHOD(jobject, VerifyOrEstablishConnection)
+(JNIEnv * env, jobject thiz, jlong commissioningWindowTimeoutSec, jobject desiredEndpointFilterJavaObject)
+{
+    chip::DeviceLayer::StackLock lock;
+    ChipLogProgress(AppServer, "CastingPlayer-JNI::VerifyOrEstablishConnection() called with a timeout of: %ld seconds",
+                    static_cast<long>(commissioningWindowTimeoutSec));
+
+    // Convert the CastingPlayer jlong to a CastingPlayer pointer
+    jclass castingPlayerClass         = env->GetObjectClass(thiz);
+    jfieldID _cppCastingPlayerFieldId = env->GetFieldID(castingPlayerClass, "_cppCastingPlayer", "J");
+    VerifyOrReturnValue(
+        _cppCastingPlayerFieldId != nullptr, nullptr,
+        ChipLogError(AppServer, "CastingPlayer-JNI::VerifyOrEstablishConnection() Warning: _cppCastingPlayerFieldId == nullptr"));
+
+    jlong _cppCastingPlayerValue  = env->GetLongField(thiz, _cppCastingPlayerFieldId);
+    CastingPlayer * castingPlayer = reinterpret_cast<CastingPlayer *>(_cppCastingPlayerValue);
+    VerifyOrReturnValue(
+        castingPlayer != nullptr, nullptr,
+        ChipLogError(AppServer, "CastingPlayer-JNI::VerifyOrEstablishConnection() Warning: castingPlayer == nullptr"));
+
+    // Create a new Java CompletableFuture
+    jclass completableFutureClass          = env->FindClass("java/util/concurrent/CompletableFuture");
+    jmethodID completableFutureConstructor = env->GetMethodID(completableFutureClass, "<init>", "()V");
+    jobject completableFutureObj           = env->NewObject(completableFutureClass, completableFutureConstructor);
+    jobject completableFutureObjGlobalRef  = env->NewGlobalRef(completableFutureObj);
+    VerifyOrReturnValue(
+        completableFutureObjGlobalRef != nullptr, nullptr,
+        ChipLogError(AppServer,
+                     "CastingPlayer-JNI::VerifyOrEstablishConnection() Warning: completableFutureObjGlobalRef == nullptr"));
+
+    ConnectCallback callback = [completableFutureObjGlobalRef](CHIP_ERROR err, CastingPlayer * playerPtr) {
+        ChipLogProgress(AppServer, "CastingPlayer-JNI::VerifyOrEstablishConnection() ConnectCallback called");
+        if (completableFutureObjGlobalRef == nullptr)
+        {
+            // Prevents an app crash at CallBooleanMethod
+            ChipLogError(AppServer,
+                         "CastingPlayer-JNI::VerifyOrEstablishConnection() ConnectCallback, Warning: completableFutureObjGlobalRef "
+                         "== nullptr");
```

Comment:
This should never happen. Can you describe why you are still seeing this?

Reply:
I am not seeing this, but thought it might be a useful check to prevent an app crash if the JNI GlobalRef is ever lost due to a programming error or some system issue.
Since it needs to be explicitly cleared with env->DeleteGlobalRef(completableFutureObjGlobalRef); it's probably low risk so I'll remove this check.